crypto/evp/p_seal.c

```diff
@@ -21,16 +21,24 @@ int EVP_SealInit(EVP_CIPHER_CTX *ctx, const EVP_CIPHER *type,
                  EVP_PKEY **pubk, int npubk)
 {
     unsigned char key[EVP_MAX_KEY_LENGTH];
-    const OSSL_PROVIDER *prov = EVP_CIPHER_provider(type);
-    OSSL_LIB_CTX *libctx = prov != NULL ? ossl_provider_libctx(prov) : NULL;
+    const OSSL_PROVIDER *prov = NULL;
+    OSSL_LIB_CTX *libctx = NULL;
     EVP_PKEY_CTX *pctx = NULL;
+    const EVP_CIPHER *cipher;
     int i, len;
     int rv = 0;
 
-    if (type) {
+    if (type != NULL) {
         EVP_CIPHER_CTX_reset(ctx);
         if (!EVP_EncryptInit_ex(ctx, type, NULL, NULL, NULL))
             return 0;
+        if ((prov = EVP_CIPHER_provider(type)) != NULL)
+            libctx = ossl_provider_libctx(prov);
+    }
+    if (libctx == NULL) {
+        cipher = EVP_CIPHER_CTX_cipher(ctx);
+        if (cipher != NULL && (prov = EVP_CIPHER_provider(type)) != NULL)
```

Comment:
Why are we assigning `prov = EVP_CIPHER_provider(type)` again; I don't see how the result would be different than when we tried it at line 35.
Also we aren't checking `type` for NULL here.